### ISSUE TITLE
RTD: Redirect pages to `.html` files

### ISSUE THREAD

**Issue by LecrisUT** [executablebooks/mystmd#984]
## Description

Navigating a static page seems to be broken when hosted on RTD. The files are there, but the path `/quickstart` does not automatically translate to `/quickstart.html`

## Proposed solution

Wish I knew :disappointed:

**Comment by mfisher87** [comment]
I'd really like to be able to deploy my MyST sites to RTD solely for the PR preview functionality. It's the best and most trustworthy implementation out there (I've read horror stories about Netlify billing people for preview sites, so won't touch it).

Are there any technical details that can be added that might help me tackle this? I have fairly low confidence in my ability to contribute code changes to this project, so any support would make a huge difference.

**Comment by choldgraf** [comment]
This is the parent issue for this one, and also worth checking out

- https://github.com/jupyter-book/mystmd/issues/1133

I agree this would be really impactful, particularly if we want to support more developer workflows since RTD is the community standard there. I wonder if this is related to the remix work that @agoose77 has been talking about for a few months. Angus do you have any ideas there?

**Comment by mfisher87** [comment]
That was quick, thanks Chris! :) I have already checked out that parent issue. This sub-issue (and #950) seems to be the core blocker; the other issues are mere inconveniences from what I can tell.

**Comment by choldgraf** [comment]
Well one thing you could try would be to see if this is fixable with this RTD setting:

https://docs.readthedocs.com/platform/latest/user-defined-redirects.html#exact-redirects

Not sure if that would work or not but worth a try

**Comment by mfisher87** [comment]
I tried the following redirects:

<img width="1986" height="766" alt="Image" src="https://github.com/user-attachments/assets/6bedf92b-35c4-403d-9b25-ede11f4d1cc2" />

None of those resolved this issue for me; you can find the deployed site here: https://geojupyter-events.readthedocs.io/

The last one is a "clean URL" redirect, which @LecrisUT reports here doesn't work with MyST: https://github.com/readthedocs/readthedocs.org/issues/11220

**Comment by mfisher87** [comment]
@LecrisUT thanks so much for opening the upstream issue! 🙇

**Comment by LecrisUT** [comment]
> [@LecrisUT](https://github.com/LecrisUT) thanks so much for opening the upstream issue! 🙇

Pleasure. Sorry I wasn't able to be more active in the ecosystem. I'm not in academia rn, but if I have time to work on some loose ends, I'll probably come back and follow up on my open issues here.

@humitos mentioned that the new react should be solving the hydration issues, but I hadn't had time to follow up on it. If anyone can pick that up, it would be great.

**Comment by mfisher87** [comment]
> [@humitos](https://github.com/humitos) mentioned that the new react should be solving the hydration issues, but I hadn't had time to follow up on it. If anyone can pick that up, it would be great.

Could you link to this? I must be looking in the wrong place :)

**Comment by LecrisUT** [comment]
> > [@humitos](https://github.com/humitos) mentioned that the new react should be solving the hydration issues, but I hadn't had time to follow up on it. If anyone can pick that up, it would be great.
> 
> Could you link to this? I must be looking in the wrong place :)

https://github.com/readthedocs/addons/issues/278#issuecomment-2067028416

And the last comment in the thread. Iirc the hydration is done in the themes repos

**Comment by ebolyen** [comment]
Hey all, I just wanted to mention that this is kind of solvable with an alternative theme:
https://github.com/jupyter-book/myst-theme/pull/530

We are currently using it for the QIIME 2 ecosystem:
https://amplicon-docs.qiime2.org/en/latest/

Downside is the myst.xrefs.json are then incorrect. 
We hack around this via a proxy which rewrites those dynamically, but it's... not good.

---

I would somewhat argue that it's a bit strange to presume a webserver would redirect a pretty-url into .html, so it might be reasonable to un-bake that assumption from the compilation step.

**Comment by choldgraf** [comment]
@ebolyen do you know how things like docusaurus do this? They have pretty URLs afaik and presumably don't run into this issue.

Also, wow I totally missed that PR. FYI at least personally I almost never look at PRs unless they're out of draft mode. Are you still pushing on that one or did you find a different solution?

**Comment by ebolyen** [comment]
I need to rebase it to catch up, but yes we're still using the compiled theme.

That said, I am kind of hoping to find a better solution w.r.t. the xrefs, but this would require the render step to be aware of pretty vs html.

Re: docusaurus, it looks like they use the same strategy as sphinx (i.e. you can have a `dirhtml` target of `foo/index.html` or `foo.html`) depending on configuration. https://docusaurus.io/docs/advanced/routing#routes-become-html-files 
They incidentally have a nice table of various "next-gen" web hosts and how they happen to configure things.

**Comment by ebolyen** [comment]
Maybe to be extra clear, it is pretty common for a server to route:
```
example.com/foo   -> /var/www/foo/index.html
```
and may or may not respond with a canonicalization of `example.com/foo/` but that usually doesn't bother anyone.

This is what I would call a "pretty url" for older static file servers like Apache and Nginx.

The newer serverless-style systems seem to be more flexible in their routing since they don't need to keep the above rules as baggage for filesystem routing, and besides, half of the page is server-side rendered on the fly anyways.

Perhaps a middle-ground would be to emit a `dirhtml` style build if configured? Then the themes can basically stay the same and it's only the rendering step that has to think about this. Does myst still crawl the served webpage, or has that improved with a static-site compatible router?

Presumably we could post-process all `<name>.html` files to `<name>/index.html` after the fact. (I might try that sometime soon actually and see if we can drop our fork of the myst theme).

**Comment by stefanv** [comment]
xref: #2178

**Comment by choldgraf** [comment]
I believe that this has been resolved by @stefanv's PR here:

- https://github.com/jupyter-book/mystmd/pull/2178

It's also released! So test it out on ReadTheDocs and let us know if / how it works! I'll close this now and we can iterate from there.

**Comment by mfisher87** [comment]
Nice! Is it released only on NPM at the moment? My RTD build uses conda-forge, and I could re-tool it, but if 1.6.0 is coming out on conda-forge soon anyway I may just wait :)

**Comment by choldgraf** [comment]
Just npm for now, I think. I'm not sure who is managing the conda forge distribution

**Comment by rowanc1** [comment]
Just released on conda. Usually takes a few hours for the branch to show up. I usually merge it.

**Comment by mfisher87** [comment]
Hooray! Thanks Rowan. I tested locally with the NPM package and it produced what it said on the tin in that PR, so I have high hopes for RTD compatibility 😁 thanks @stefanv !

**Comment by weiji14** [comment]
Woohoo, thanks everyone! I've [tested `jupyter-book=2.0.0b2`](https://github.com/weiji14/cog3pio/pull/53) on Readthedocs PR preview, and can confirm that clean urls work!

**Comment by lsetiawan** [comment]
Yes! Thank you so much everyone!! I can also [confirm this](https://github.com/seafloor-geodesy/gnatss-workshop/pull/25). For anyone wanting a conda-forge + pypi + RTD solution with pixi, do check out the repo I'm working on for a [workshop](https://github.com/seafloor-geodesy/gnatss-workshop) here at the UW eScience Institute 😄. Definitely an experiment moving from our typical jupyterbook 1.0 for workshops to 2.0.

**Comment by mfisher87** [comment]
Hell yeah pixi 😁

**Comment by mfisher87** [comment]
My ReadTheDocs builds are working great. Maybe we should add a documentation page. I have "single version" mode turned on to avoid dealing with `BASE_URL` (I think we can improve on this, as e.g. Quarto "just works" with language and version e.g. `/en/stable` in the URL) and "clean URL" redirects set up.

**Comment by choldgraf** [comment]
@mfisher87 want to add a little `Hosting on ReadTheDocs` page? I'm happy to review a PR.

For what it's worth we got the language / version redirecting working OK with this configuration:

https://github.com/jupyter-book/jupyter-book/blob/next/.readthedocs.yaml

you can see it here!

https://jupyterbook.org/en/next

**Comment by mfisher87** [comment]
Absolutely! I don't think I'll get to it today, but I'd be happy to. I opened https://github.com/jupyter-book/mystmd/issues/2196 -- would you mind assigning to me and giving feedback on the proposal?
